corehq/form_processor/backends/sql/dbaccessors.py

```diff
@@ -586,13 +586,30 @@ def save_deprecated_form(form):
         FormAccessorSQL.save_new_form(form)
 
     @staticmethod
-    @transaction.atomic
-    def update_form_problem_and_state(form):
-        with get_cursor(XFormInstanceSQL) as cursor:
-            cursor.execute(
-                'SELECT update_form_problem_and_state(%s, %s, %s)',
-                [form.form_id, form.problem, form.state]
-            )
+    def update_form(form):
+        from corehq.form_processor.change_publishers import publish_form_saved
+        from corehq.sql_db.util import get_db_alias_for_partitioned_doc
+        assert form.is_saved(), "this method doesn't support creating unsaved forms"
+
+        db_name = get_db_alias_for_partitioned_doc(form.form_id)
+        if form.orig_id:
+            old_db_name = get_db_alias_for_partitioned_doc(form.orig_id)
+            assert old_db_name == db_name, "this method doesn't support moving the form to new db"
+
+        if form.form_id_updated():
+            attachments = form.original_attachments
+            operations = form.original_operations
+            with transaction.atomic(db_name):
+                form.save(using=db_name)
+                for a in attachments:
+                    a.form = form
+                    a.save(using=db_name)
+                for op in operations:
+                    op.form = form
+                    op.save(using=db_name)
+        else:
+            form.save(using=db_name)
```

Comment:
if you merge master you can remove the `using=db_name` from the save call since that is now automatic